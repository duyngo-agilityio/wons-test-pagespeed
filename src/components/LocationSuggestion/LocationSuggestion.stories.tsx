import { useState, useEffect, ChangeEvent } from 'react';
import type { Meta, StoryObj } from '@storybook/react';

// Libraries
import { useDebounce } from 'use-debounce';

// Icons
import { HiLocationMarker } from 'react-icons/hi';

// Components
import LocationSuggestion from './index';
import { Input } from '../common';

// Types
interface LocationItem {
  properties: {
    formatted: string;
    name: string;
  };
}

const meta: Meta<typeof LocationSuggestion> = {
  title: 'Components/LocationSuggestion',
  component: LocationSuggestion,
  parameters: {
    layout: 'centered',
    controls: { expanded: true },
  },
};

export default meta;
type Story = StoryObj<typeof LocationSuggestion>;

// Mocked data
const mockLocationResponse = {
  features: [
    { properties: { formatted: 'New York, NY', name: 'New York' } },
    { properties: { formatted: 'Los Angeles, CA', name: 'Los Angeles' } },
  ],
};

const mockGetLocationSuggestion = async (_keyword: string) => {
  return mockLocationResponse;
};

const BasicUsage = () => {
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const [searchAddress, setSearchAddress] = useState<string>();
  const [locationsSuggestion, setLocationsSuggestion] = useState<
    LocationItem[]
  >([]);

  // Debounce searchAddress with a delay of 300ms
  const [debouncedSearchAddress] = useDebounce(searchAddress, 300);

  useEffect(() => {
    const handleFetchData = async (keyword: string) => {
      const locationResponse = await mockGetLocationSuggestion(keyword);

      setLocationsSuggestion(
        locationResponse?.features ? locationResponse.features : [],
      );
    };

    if (debouncedSearchAddress) {
      handleFetchData(debouncedSearchAddress);
    } else {
      setLocationsSuggestion([]);
    }
  }, [debouncedSearchAddress]);

  const handleClickSuggestion = (value: string) => {
    setSearchAddress(value);
    setLocationsSuggestion([]);
    setIsOpen(false);
  };

  const handleOnChangeSearch = (event: ChangeEvent<HTMLInputElement>) => {
    setIsOpen(true);

    const value = event.target.value;
    setSearchAddress(value);
  };

  return (
    <div className="relative flex-1 h-[150px]">
      <Input
        placeholder="Enter your address"
        onChange={handleOnChangeSearch}
        endContent={<HiLocationMarker />}
        value={searchAddress}
      />
      {locationsSuggestion.length !== 0 && isOpen && (
        <div className="shadow-md p-3 absolute z-50 bg-white dark:bg-gray-400 w-full">
          {locationsSuggestion.map(({ properties }: LocationItem) => {
            const { formatted, name } = properties;

            return (
              <LocationSuggestion
                formatted={formatted}
                key={name}
                onClickSuggestion={handleClickSuggestion}
              />
            );
          })}
        </div>
      )}
    </div>
  );
};

export const Default: Story = {
  render: () => <BasicUsage />,
};
