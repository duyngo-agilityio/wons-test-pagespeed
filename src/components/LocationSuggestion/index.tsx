import { memo } from 'react';
import { HiLocationMarker } from 'react-icons/hi';

import { Text } from '@/components';

interface LocationSuggestionProps {
  onClickSuggestion: (value: string) => void;
  formatted: string;
}

const LocationSuggestion = ({
  onClickSuggestion,
  formatted,
}: LocationSuggestionProps) => {
  const handleClickSuggestion = () => {
    onClickSuggestion(formatted);
  };

  return (
    <div
      className="flex w-full text-left py-[10px] px-[4px] gap-4 items-center relative cursor-pointer dark:text-white/80"
      onClick={handleClickSuggestion}
    >
      <div className="w-[16px] h-[16px]" data-testid="location-icon">
        <HiLocationMarker />
      </div>

      <Text text={formatted} className="dark:text-white/80" />
    </div>
  );
};

export default memo(LocationSuggestion);
