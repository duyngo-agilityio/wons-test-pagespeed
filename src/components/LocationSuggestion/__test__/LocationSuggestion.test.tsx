import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

// Components
import LocationSuggestion from '../index';

describe('LocationSuggestion Component', () => {
  const mockOnClickSuggestion = jest.fn();
  const formatted = '123 Test Street, Test City';

  const setup = () =>
    render(
      <LocationSuggestion
        onClickSuggestion={mockOnClickSuggestion}
        formatted={formatted}
      />,
    );

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with location text and icon', () => {
    setup();

    expect(screen.getByText(formatted)).toBeInTheDocument();
    expect(screen.getByTestId('location-icon')).toBeInTheDocument();
  });

  it('calls onClickSuggestion with formatted text when clicked', () => {
    setup();

    const suggestionDiv = screen.getByText(formatted);
    fireEvent.click(suggestionDiv);

    expect(mockOnClickSuggestion).toHaveBeenCalledWith(formatted);
  });

  it('matches snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
});
