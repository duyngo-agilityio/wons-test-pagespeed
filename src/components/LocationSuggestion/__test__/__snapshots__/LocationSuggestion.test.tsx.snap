// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationSuggestion Component matches snapshot 1`] = `
<div>
  <div
    class="flex w-full text-left py-[10px] px-[4px] gap-4 items-center relative cursor-pointer dark:text-white/80"
  >
    <div
      class="w-[16px] h-[16px]"
      data-testid="location-icon"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5.05 4.05a7 7 0 119.9 9.9L10 18.9l-4.95-4.95a7 7 0 010-9.9zM10 11a2 2 0 100-4 2 2 0 000 4z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <p
      class="font-dm-sans text-blue-800 dark:text-white text-md dark:text-white/80"
    >
      123 Test Street, Test City
    </p>
  </div>
</div>
`;
